src/data/multiset/fintype.lean

```diff
@@ -0,0 +1,222 @@
+/-
+Copyright (c) 2022 Kyle Miller. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Kyle Miller
+-/
+import data.fintype.card
+import algebra.big_operators
+
+/-!
+# Multiset coercion to type
+
+This module defines a `has_coe_to_sort` instance for multisets and gives it a `fintype` instance.
+It also defines `multiset.to_enum_finset`, which is another way to enumerate the elements of
+a multiset. These coercions and definitions make it easier to sum over multisets using existing
+`finset` theory.
+
+## Main definitions
+
+* A coercion from `m : multiset α` to a `Type*`. For `x : m`, then there is a coercion `↑x : α`,
+  and `x.2` is a term of `fin (m.count x)`. The second component is what ensures each term appears
+  with the correct multiplicity. Note that this coercion requires `decidable_eq α` due to
+  `multiset.count`.
+* `multiset.to_enum_finset` is a `finset` version of this.
+* `multiset.coe_embedding` is the embedding `m ↪ α × ℕ`, whose first component is the coercion
+  and whose second component enumerates elements with multiplicity.
+* `multiset.coe_equiv` is the equivalence `m ≃ m.to_enum_finset`.
+
+## Tags
+
+multiset enumeration
+-/
+
+open_locale big_operators
+
+variables {α : Type*} [decidable_eq α] {m : multiset α}
+
+/-- Create a type that has the same number of elements as the multiset.
+Terms of this type are triples `⟨x, ⟨i, h⟩⟩` where `x : α`, `i : ℕ`, and `h : i < m.count x`.
+This way repeated elements of a multiset appear multiple times with different values of `i`. -/
+instance : has_coe_to_sort (multiset α) Type* :=
+⟨λ m, Σ (x : α), fin (m.count x)⟩
+
+/-- As a convenience, there is a coercion from `m : Type*` to `α` by projecting onto the first
+component. -/
+instance multiset.has_coe_to_sort.has_coe : has_coe m α := ⟨λ x, x.1⟩
+
+@[simp] lemma multiset.fst_coe_eq_coe {x : m} : x.1 = x := rfl
+
+@[simp] lemma multiset.coe_eq {x y : m} : (x : α) = (y : α) ↔ x.1 = y.1 :=
+by { cases x, cases y, refl }
+
+@[simp] lemma multiset.coe_mk {x : α} {i : fin (m.count x)} : ↑(⟨x, i⟩ : m) = x := rfl
+
+@[simp] lemma multiset.coe_mem {x : m} : ↑x ∈ m := multiset.count_pos.mp (pos_of_gt x.2.2)
+
+@[simp] protected lemma multiset.forall_coe (p : m → Prop) :
+  (∀ (x : m), p x) ↔ ∀ (x : α) (i : ℕ) (h : i < m.count x), p ⟨x, i, h⟩ :=
```

Comment:
I think this would be better if the quantifier used `fin (m.count x)` on the right, so that the user can do the two rewrites separately

Reply:
Done. It would be nice if `fin.exists_iff` and `fin.forall_iff` were simp lemmas, but I'm not going to embark on that project here.